mint-auditor/src/db/models/audited_mint.rs

```diff
@@ -0,0 +1,696 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{
+    db::{schema::audited_mints, transaction, Conn, Counters, GnosisSafeDeposit, MintTx},
+    gnosis::{AuditedSafeConfig, GnosisSafeConfig},
+    Error,
+};
+use diesel::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// Diesel model for the `audited_mints` table.
+/// This stores audit data linking MintTxs with matching GnosisSafeDeposits.
+#[derive(Clone, Debug, Deserialize, Eq, Hash, Insertable, PartialEq, Queryable, Serialize)]
+pub struct AuditedMint {
+    /// Id (required to keep Diesel happy).
+    pub id: Option<i32>,
+
+    /// Id pointing to the MintTx table.
+    pub mint_tx_id: i32,
+
+    /// Id pointing to the GnosisSafeDeposit table.
+    pub gnosis_safe_deposit_id: i32,
+}
+
+impl AuditedMint {
+    /// Attempt to find a matching [MintTx] for a given [GnosisSafeDeposit], and
+    /// if successful return the [MintTx] and record the match in the
+    /// database. Note that each [MintTx] can be matched to at most one
+    /// [GnosisSafeDeposit], so calling this repeatedly over the same
+    /// deposit will fail.
+    pub fn attempt_match_deposit_with_mint(
+        deposit: &GnosisSafeDeposit,
+        config: &AuditedSafeConfig,
+        conn: &Conn,
+    ) -> Result<MintTx, Error> {
+        // We only operate on objects that were saved to the database.
+        let deposit_id = deposit.id().ok_or(Error::ObjectNotSaved)?;
+
+        // The deposit safe needs to match the audited safe configuration.
+        // This shouldn't happen and indicates misuse of this function.
+        if deposit.safe_addr() != &config.safe_addr {
+            return Err(Error::Other(format!(
+                "Gnosis safe deposit addr {} does not match audited safe addr {}",
+                deposit.safe_addr(),
+                config.safe_addr
+            )));
+        }
+
+        let result = transaction(conn, |conn| {
+            // Currently we only support 1:1 mapping between deposits and mints, so ensure
+            // that there isn't already a match for this deposit.
+            let existing_match = audited_mints::table
+                .filter(audited_mints::gnosis_safe_deposit_id.eq(deposit_id))
+                .first::<AuditedMint>(conn)
+                .optional()?;
+            if let Some(existing_match) = existing_match {
+                return Err(Error::AlreadyExists(format!(
+                    "GnosisSafeDeposit id={} already matched with mint_tx_id={}",
```

Comment:
I think updating the table to use `nonce`/`eth_tx_hash` would be somewhat less efficient both in terms of storage, and in terms of query speed. In reality it is unlikely to matter since the amount of mints we expect to have is fairly small. Even at a few a month this is still an insignificant amount of data for a long time.

I'll see how to change the query to join in the nonce and hash so that the error message is a bit more useful.